### PR TITLE
Added Guides and Node Flexibility Tutorial.

### PR DESCRIPTION
**Pull request by willnationsdev**
This creates a supplemental introduction to Godot that should work alongside the step-by-step tutorial in order to quickly get someone familiarized with Godot's systems and link them to relevant parts of the documentation.

The step-by-step tutorial specializes in getting people's feet wet with an actual project and teaches them Godot's basics along the way. This guide focuses on explaining Godot's most important concepts to users at a high level. If it gets too in-depth, then it favors directing the user to other, more detailed parts of the docs. For example, as I was going into detail about node flexibility, there was too much information, so I outsourced it to an entirely new "Node Flexibility" tutorial section.

It has the luxury of not needing to be a tutorial itself. This enables it to move faster, cover more topics, and get people kickstarted more quickly. It is especially useful to people who are already familiar with programming, due to the speed with which it covers important topics. However, it has been organized with the assumption that the reader knows next-to-nothing about programming.

Since I noticed there was a Unity-to-Godot guide in the `getting_started/editor` section as well, I figured it was in the spirit of things to create a dedicated subsection for "guides". We could then conceivably add more guides on getting started with Godot that are tailored to different audiences.

I understand that this is a huge update, but it is ultimately one full-length document, not an entire series like I did before. It was important for me to put it all together in one PR as many parts of this were actually rearranged as I worked on it (even just hours before this submission). I don't expect this to get merged soon, given the amount of content that needs to be reviewed.

A few caveats:

1. I can't compile this at the moment because the `getting_started/animation/introduction_2d` tutorial has some problems with images / image directives.
2. I still need to compress the many images associated with this PR.

Please let me know what you think. Thanks.

### CODE REVIEW

**Comment by RaTcHeT302** [comment]
i read the text, and i think it's too verbose, plus there's a lot of knowledge which you need to have beforehand in order to really understand anything from it

i don't think this is very helpful for new users as it is currently

it sounds very unfocused, and i don't feel like this will help anyone, if anything it looks a bit scary, and it would probably discourage new users 

i think this needs to be trimmed down, to what's truly essential, and it should only teach one thing, or one concept, at a time, but instead, currently it feels like a giant information dump, which is not very useful for someone who's getting started out

this doesn't feel like it was written for new users at all, i think the language should also be simplified, to an elementary school \ high school level, so no any fancy words, just plain and clear english



this adds nothing of value, it's filler, it can be removed - text like this, wastes my time - most people are not going to read this whole thing, most users are going to scan the text, skip the boring parts, or anything which they don't understand, and then they'll move on

overall, there's too much fluff text, and there are too many terms, which you have to be aware of beforehand

i think the information overload is too overwhelming, but i don't know what this tutorial actually looks like in godot, i only looked at the text in the commits

**Comment by RaTcHeT302** [comment]
i think instead of trying to over explain everything, all at once, every concept, should be broken up into it's own guides

example

setup
- how to make a main loop

variables
- how to create a variable
- how to modify a variable
- how to keep track of a variable

ui
- whatever ui stuff

i don't know the actual topics, but generally i would focus on explaining one thing at one time

i would personally avoid telling the readers, to remember things in the future, to me this says, that the guide's structure is flawed, the logic is too spread out, it's trying to teach WAAAY too much at once


i think this would be more helpful, if it had a more, "get it done", type of style to it

new users, are not gonna remember everything from this, building such knowledge takes time, you are throwing too much at once, i think setting up the main loop, is probably complicated enough as t is, for someone who is starting out

instead this needs to be spread out, into multiple guides, which teach a specific concept

that way users, can refer to them, in any order they'd like to, and they can choose, what they want to learn, it needs to be decoupled into separate topics

**Comment by garyritchie** [comment]
In my opinion this document serves as an excellent [explanation](https://www.divio.com/blog/documentation/) for newcomers to Godot. I have a working knowledge of JavaScript and C# so I can appreciate both the brevity _and_ linking to further information.

I will be referring to this as I continue to learn Godot while starting my first project. Thanks Will!

**Comment by willnationsdev** [comment]
@RaTcHeT302 

> i read the text, and i think it's too verbose, plus there's a lot of knowledge which you need to have beforehand in order to really understand anything from it
> i don't think this is very helpful for new users as it is currently
> this doesn't feel like it was written for new users at all,
> i think the language should also be simplified, to an elementary school \ high school level, so no any fancy words, just plain and clear english
> this adds nothing of value, it's filler, it can be removed - text like this, wastes my time - most people are not going to read this whole thing, most users are going to scan the text, skip the boring parts, or anything which they don't understand, and then they'll move on
> overall, there's too much fluff text, and there are too many terms, which you have to study by themselves

Perhaps the target audience of this should be adjusted(?). The pain goal of this document was to have a comparatively brief overview of the essential Godot features that are relevant to people coming to the engine for the first time, and to allow them to move quickly through the document, skimming if necessary, and link them to additional resources for in-depth topics.

Because newcomers to Godot have a wide range of experience levels though, I have tried to take any jargon or unfamiliar words and either define them myself, in the context of Godot, or link them to external authorities (such as Wikipedia) to help explain the term(s).


> i think this needs to be trimmed down, to what's truly essential

If you believe certain sections to not be critical parts of understanding the best way to use Godot Engine to its fullest extent, then could you please elaborate on which sections those are specifically? The topics covered thus far are, at least in my opinion, absolutely essential things to know in order to adequately build medium-to-large projects in Godot Engine.

> it sounds very unfocused...if anything it looks a bit scary, and it would probably discourage new users
> it should only teach one thing, or one concept, at a time
> instead, currently it feels like a giant information dump, which is not very useful for someone who's getting started out
> i think the information overload is too overwhelming
> i think instead of trying to over explain everything, all at once, every concept, should be broken up into its own guide

If we were to do this, then it would be functionally identical to the myriad of `tutorial` section documents, but without all of the detail. This kind of documentation's entire purpose is to lightly cover a wide breadth of topics in as little detail as possible, while still communicating critical information, and then linking to additional information as needed. Being a giant information dump for people coming into Godot is exactly what it is meant to be. Rather than over explaining things, I barely touched a huge variety of topics and purposefully did not go into details with actual code examples, use cases, demonstrations, etc. That would be more useful in an in-depth tutorial.

> i don't know what this tutorial actually looks like in godot, i only looked at the text in the commits

The only part of it that would look significantly different is the Node Flexibility Tutorial which has tons and tons of images included with it.

> i would personally avoid telling the readers, to remember things in the future, to me this says, that the guide's structure is flawed, the logic is too spread out, it's trying to teach WAAAY too much at once

That's a fair point, and I mostly meant to refer to those "tree features" in the Node Flexibility tutorial anyway, which ended up getting outsourced to another page, so it seems like a good idea to remove all cross-section references altogether in the page. Especially if it helps keep things less scary for readers.

> new users, are not gonna remember everything from this

That's fine. If someone wants to read through it multiple times over different instances to let information sink in, that's great. And if they just want to use it as a reference, jumping to the relevant section so that they can find the links to other relevant documentation, then that's good too.

> for someone who is starting out

Yeah, I definitely think the target audience was misinterpreted/miscommunicated. By "knows next-to-nothing about programming", I meant that there would be users who might be familiar with other game engines, but who don't really work with code (GameMaker/Studio, Construct, Stencyl, etc.).

This document is tailored towards people who just want an understanding of how Godot works or how Godot differentiates from other programming paradigms they have experienced (even if they aren't familiar with game programming, text-based programming, or Godot specifically).

> i would have a giant list of topics, each topic would be self contained ideally if possible, so that users, could learn to do stuff from them, in any order if they wanted to
> redundancy would be fine, it helps users build familiarity
> this would benefit new users, and experienced users, who would like a quick reminder

Honestly, this is exactly what the `tutorial` section already does. It sounds like you just want more topics to be there so that Google searches can SEO their way into the perfect community-maintained tutorial to fit their needs.

To do that, people just need to learn how to add content to the docs. XD

> the current guide is nothing more than a glorified blog post, i wouldn't read it personally, i would look for something else - it's not presented in a way, which makes me want to sit down and read the whole thing

Can you elaborate on how it makes you feel like it's a blog post, or why, even in that case, a blog post isn't informative or helpful?

If I understand what you mean, you are saying that it is too wordy, doesn't slow down to carefully deconstruct each topic, and needs to separate topics to be more isolated. Aside from the wordiness aspect though, everything else is actually purposeful: description brevity, topic breadth, fast pace, all in one comparatively concise document. It could've been a tutorial series with several thousands of lines of content, including dozens of images. But that is not its intent.

The wordiness though is definitely something that gets in the way of the pace, so I'd like to fix that if I can.

> the time it takes me to google, is shorter, than the time it would take me, to decipher the whole guide, or to extract key pieces of info from it - it's not immediate - it's too long to be of any use

See above.

> if you'd like a quick reminder

This is for a dedicated tutorial on a single topic or for API doc referencing. Not the purpose of this document.

> it's too fluffy for a new user

This is good feedback. I mean, I'll try to proofread it again, but writers seldom are effective at proofreading their own work. I could just as easily have *actually* [made it a blog post](https://willnationsdev.wordpress.com/2018/04/07/godots-node-system-part-3-engine-comparisons/), if it's on GitHub in the official docs, people can do line-by-line comments, pointing out exactly where there are issues. Just telling me that it is too fluffy is useful, but I can't guarantee that I'll actually fix everything that is too fluffy unless someone points out specific instances.

---

I really appreciate you taking the time to write such a detailed response. I will make sure to back through it and see if I can't make it more compact/direct. Hopefully the target audience and intended style of the guide is more clear at this point.

File: getting_started/guides/godot_essentials_guide.rst [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+
+    This extends to scenes. They can have embedded resources such as
+    meshes, materials, shaders, particle effects, etc.
+    
+    However, Godot will try to save on memory allocations unless directed
+    otherwise. If you duplicate a scene, or if you copy/paste a
+    sub-resource, then it will NOT create a new resource. Instead, it
+    makes both scenes reference the same sub-resource.
+
+    To force Godot to create a new sub-resource, do X.
```
Comment (willnationsdev) [inline getting_started/guides/godot_essentials_guide.rst]: Oh whoops. Looks like I left a todo in here.

File: getting_started/guides/godot_essentials_guide.rst [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+   easy-to-use
+   :ref:`plugin API <class_EditorPlugin>`. For one-off edits, you can also
+   write :ref:`standalone editor scripts <class_EditorScript>`.
+   For more information about how to create plugins,
+   :ref:`click here <doc_making_plugins>`.
+
+3. There's no need to learn a domain-specific language for plugins.
+   Instead, Godot enables you to re-use your game UI knowledge.
+
+4. While writing a plugin, you can also add
```
Comment (willnationsdev) [inline getting_started/guides/godot_essentials_guide.rst]: Maybe also add a 5th point that elaborates briefly and links to other *Plugin subtypes such as importers and the like. Otherwise I only touch on the InspectorPlugins and nothing else.

File: getting_started/workflow/best_practices/godot_notifications.rst [hunk 3]
```diff
@@ -297,3 +297,92 @@ nodes that one might create at runtime.
...
+    extends Node
+    func _ready():
+        add_to_group("f_game_over")
+    func _game_over():
+        pass # Do something
+    
+    # Notifier
+    extends Node
+    func game_over():
+        get_tree().call_group("f_game_over", "_game_over")
```
Comment (Feniks-Gaming) [inline getting_started/workflow/best_practices/godot_notifications.rst]: You are not adhering to the use of white space suggested by a style guide. 

There should be 2 blank lines between each function.

**Comment by mhilbrunner** [comment]
There is some useful information here, but where and how we add it may depend on the outcome of the proposal for redesigning the docs for Godot 4.0: https://github.com/godotengine/godot-docs/issues/3390

So it may take a while :)

**Comment by willnationsdev** [comment]
@mhilbrunner Yeah, I kinda figured. I still need to update it to fix the stuff that's been mentioned (and probably update code samples to conform to style guide - haven't checked that yet).

File: getting_started/workflow/best_practices/godot_notifications.rst [hunk 4]
```diff
@@ -297,3 +297,92 @@ nodes that one might create at runtime.
...
+world. Or, if you already have a subset of nodes in mind, you can use
+:ref:`SceneTree.notify_group(what) <class_SceneTree_method_notify_group>`
+to similar effect, but only targeting nodes in a group. Any node that
+implements the :ref:`Node._notification <class_Node_method__notification>`
+method will be able to catch and react to those notifications.
+
+Another option is to create your own custom "notification callbacks". You
+can use :ref:`Node.propagate_call <class_Node_method_propagate_call>` on
+the SceneTree's root node to trigger a method call on every node that
+implements the method. Or, if you already havea a subset of nodes in mind,
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/godot_notifications.rst:322]: `havea` -> `have a`

File: getting_started/workflow/best_practices/godot_notifications.rst [hunk 5]
```diff
@@ -297,3 +297,92 @@ nodes that one might create at runtime.
...
+you can use :ref:`SceneTree.group_call <class_SceneTree_method_call_group>`.
+Naturally, operating with groups is going to be much less expensive.
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/godot_notifications.rst:324]: This isn't necessarily true. Should elaborate on details a smidge.

File: getting_started/guides/godot_essentials_guide.rst [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+
+1. build nodes into a tree.
+2. subdivide a tree into smaller trees.
+3. attach trees to other trees as children.
+4. reorganize the nodes within a tree to produce a new tree.
+5. masquerade a tree as a node.
+6. create and delete entire trees.
+7. move nodes between trees.
+
+We'll refer to these later in this guide as "tree features".
```
Comment (willnationsdev) [inline getting_started/guides/godot_essentials_guide.rst:155]: Remove this, as I never actually get around to doing it.

File: getting_started/guides/godot_essentials_guide.rst [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+Godot runs on (desktop, web, mobile). It also handles iterative processes like
+frame calculations and checking for input.
+
+MainLoop is then inherited by the :ref:`SceneTree <class_SceneTree>` class.
+The SceneTree is your actual Godot game instance. It builds on top of MainLoop
+to enable users to manage a world of content, change it, organize it, and help
+it communicate over networks. To be more specific, SceneTree manages a tree
+hierarchy of :ref:`Nodes <class_Node>`.
+
+Nodes, Node trees
```
Comment (willnationsdev) [inline getting_started/guides/godot_essentials_guide.rst:138]: Nodes and Node Trees

File: getting_started/guides/godot_essentials_guide.rst [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+
+  For more information and examples of signals, see the relevant documentation
+  in 
+  :ref:`GDScript <doc_signals>`, :ref:`C# <doc_c_sharp_differences>`,
+  :ref:`VisualScript <doc_getting_started_visual_script>`, and
+  :ref:`NativeScript C++ <doc_gdnative_cpp_example>` as well as how to
+  :ref:`instance scenes via signals <doc_instancing_with_signals>`.
+
+  For more practical information on when to use signals, see the best practices
+  documentation on :ref:`scene organization <doc_scene_organization>`.
```
Comment (willnationsdev) [inline getting_started/guides/godot_essentials_guide.rst:423]: Add a section on the `owner` variable for intra-scene communication.

File: getting_started/guides/godot_essentials_guide.rst [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+  node, it automatically deletes its children beforehand. And when
+  each of those children receive the command to delete themselves, then they
+  too will delete their children beforehand. This recursive process continues
+  until the leaf nodes, with no children, finally delete themselves. The node
+  tree systematically deletes itself from the bottom up!
+
+  Nodes in particular also have a
+  :ref:`queue_free() <class_Node_method_queue_free>` method. This delays
+  their deletion until the next idle frame to reduce any discernible
+  lag.
```
Comment (willnationsdev) [inline getting_started/guides/godot_essentials_guide.rst:469]: Add link to getting started docs about freeing nodes, which give info about `is_instance_valid()` and `is_queued_for_deletion()`, etc.

File: getting_started/guides/godot_essentials_guide.rst (hunk 1 above)
Comment (willnationsdev) [reply getting_started/guides/godot_essentials_guide.rst:526]: Iirc, you use the "Local to scene" checkbox to mark that a resource should create a new instance when the scene itself is copied. Then there's also a "Make Unique" operation that forces the current instance to *become* local to the current scene if it isn't already. Need to double check this info though.

**Comment by NathanLovato** [comment]
Just getting to this PR now. I think it's a useful resource for developers to get up to speed with Godot. For users with little programming experience, it'll likely break their expectations. It's not so much about the language as it is about the content's nature.

The getting started section is bound for a rewrite that'll feature two learning paths: one for users who still have limited programming experience, and one for developers. This intro guide still needs planning but I think your work fits there. Would you be ok if we reorganized the content to fit a new getting started series for devs?

**Comment by NathanLovato** [comment]
I'll share a detailed proposal soon for the getting started section, and it'll probably take some time as it'll need quite a bit of input from new users in the target audiences.

**Comment by willnationsdev** [comment]
@NathanLovato 

> Would you be ok if we reorganized the content to fit a new getting started series for devs?

Not at all. However you see fit to use this work to improve the documentation experience for those getting started with Godot. It's specifically meant to be a brief overview of the way the whole thing works, so it's a great fit for experienced devs.

**Comment by NathanLovato** [comment]
Getting back to this, I've got a few questions and requests:

1. In the future, could you make one PR per page or per big change? There's a need for edits in about every page you touched in this PR, making it cumbersome to use the PR review features.
2. Can I go ahead and pull your changes, edit and split?

I won't dive into the details or changes I'd request, because it'd touch every bit of the article to the point it wouldn't be the same thing anymore. In general, I think:

1. Its cognitive load would be too high for most users.
2. It mixes high-level abstractions all users need to know about with technical details most users won't ever deal with (servers and rids, memory).
3. I'd split the page into multiple and change the order. The article feels like a juxtaposition of concepts from an outside view, the way it's currently presented.

I'm saying that trying to put myself in the users' shoes and discovering a new technology. I can see how you thinking in branching ways, but we need to chop and bundle the info in a way that's as easy to digest as possible for users (note there are quite a few heavy pages in the docs, especially those written by experienced software engineers).

**Comment by willnationsdev** [comment]
@NathanLovato 

> In the future, could you make one PR per page or per big change? There's a need for edits in about every page you touched in this PR, making it cumbersome to use the PR review features.

Almost every bit of content in the PR is *new* content. In such a case, where I'm guessing one large article would instead become one new subsection of the docs with many small articles for each segment, how should such a thing be done? One PR for each page, but where each page is dependent on preceding pages being present to make any sense? Just not sure how it'd work in terms of merging/syncing the actual content, in case some of them need to be edited/changed such that they no longer fit together in terms of flow.

> Can I go ahead and pull your changes, edit and split? Or could you split the PR yourself, so we can have you as the original commit authors?

Yeah, I can split it up myself. As requested above, how best would you like such a PR to be split up? If I split it up by section/topic, I'd probably end up with about, idk, 10-15 PRs being spammed in godot-docs.

> Its cognitive load would be too high for most users.


You mean in terms of word choice / style, etc.? The topics at hand are targeted at getting experienced users up-to-speed more quickly, so I don't think you're talking about the level of detail it goes into (since it barely touches on topics before giving links and jumping to the next one).

> It mixes high-level abstractions all users need to know about with technical details most users won't ever deal with (servers and rids, memory).

This is on purpose. An experienced developer new to the engine needs to know what the high level abstractions are, but then they also need to know what options they have to improve the performance. How does the engine work at a high level? At a low level? What is the relationship in between? How can I slip into either one or craft such a relationship on my own? This all stuff that an experienced user will, in my experience, want to know.

What's important though is not spending too much time on the details and instead linking to more detailed articles.

If you think there's a need to reduce the amount of detail it goes into further, then that would make sense to me, but I don't think it should completely leave the information out either.

> I'd split the page into multiple and change the order. The article feels like a juxtaposition of concepts from an outside view, the way it's currently presented.

That's fine. I'll try to break it down into concrete pages and submit them as new independent PRs.

Is there a particular area of the docs I should make the files in that you would prefer? Some directory where they can be arranged as part of a series, I would think?

**Comment by NathanLovato** [comment]
The cognitive load is about several aspects: the wording that could sometimes be more accessible, the order and link between sections, but more importantly the overwhelming number of concepts covered in one place. In particular, there are dozens of links to even more concepts. Also, many details like the references to `tree feature #x`, which asks of readers to remember something while trying to process new information. You may be expecting users to use the guide as a reference they can revisit later but they'll most likely stop reading altogether as soon as the load becomes too high. That is, two or three sections in.

I would put way fewer explanations in this article and link at most to one or two pages per section that cover one precise concept at a time, in details.

When it comes to making multiple PRs, I was mostly referring to the fact you wrote two new articles. I'd recommend making at least two, one for the node flexibility page and one for the other branched from it. You just need to write something like `needs #x merged first` in your PR for reviewers to know to check the other first.

<hr>

Here are some more actionable recommendations:

1. Changing the title to stress it's intended for experienced developers. E.g. `Overview of how Godot works for professional developers`. Something like that.
2. Stress below the title the article is intended for experienced developers. You can optionally list several pre-requisites. Sounds redundant with the title but the page is not for most newcomers reading the docs yet will likely be in the getting started series so we should ensure only the intended audience reads it. I'd even add a note that links to a more beginner-friendly page.
3. As the page covers many topics, try to summarize what users will gain from the article and learn more precisely at the top. The current intro suggests the page is going to teach you *how* to build a game in Godot. I can tell you from experience many readers will expect instructions in this case, including professional software developers. Instead, it introduces concepts and gives you an overview of important features to get some sense of Godot's design.
4. Stay under 2000 words, if possible way under that. It's supposed to be an intro to some essential concepts. At this stage, you can consider your reader is only discovering the engine and everything you say is completely new to them and the details of how it works don't matter yet. If you feel you need to provide details, write them in another page and link to it.
5. The part about scenes is going to be redundant with the new getting started series. It's also different in style than the rest of the guide.
6. When referring to info previously exposed in the document, state the feature or concept's name again.
7. There's no misc/ section anymore. Node flexibility could go to best practices maybe?

That's quite a lot already. If you need anything, I'm happy to help.

**Comment by willnationsdev** [comment]
@NathanLovato Thanks for the feedback! I'm going to re-read the updated getting started section (cause it's been a *long* time since I've looked at it), and try to see where the information from this really fits in. I'll try to split things up a lot more effectively, and will submit more PRs associated with the content that are tighter in scope.

**Comment by NathanLovato** [comment]
The getting started section didn't get a rewrite yet. Here's the plan so far (waiting for approval from Akien): #4035

**Comment by willnationsdev** [comment]
Ah! Good to know. Well, I'll just pick where I think it best to put things, and you can just tell me in individual PRs if some piece of information should be moved I guess.

**Comment by mhilbrunner** [comment]
This has been lingering for a while, there are conflicts, but mostly the discussion above outlined why this is hard to merge as-is. There's some good stuff here, and some may be good to have in separate smaller, more focused PRs.

The docs however have changed a lot since the initial PR.

I'm going to close this one, and still thanks for all the work that went into it!

**Comment by willnationsdev** [comment]
@mhilbrunner That makes sense. I may have to split it up later, when I have more time.
